src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11767,62 +11767,62 @@ private bool CanStartConsequenceExpression()
+                // while 'a?.b!' should be a suppression-expr containing a cond-access 'a?.b'.
+                using var beforeSuppressionsResetPoint = GetDisposableResetPoint(resetOnDispose: false);
+                while (this.CurrentToken.Kind == SyntaxKind.ExclamationToken)
+                    expr = _syntaxFactory.PostfixUnaryExpression(SyntaxKind.SuppressNullableWarningExpression, expr, EatToken());
 
                 switch (this.CurrentToken.Kind)
                 {
+                    // a?.b(...)
                     case SyntaxKind.OpenParenToken:
-                        expr = _syntaxFactory.InvocationExpression(expr, this.ParseParenthesizedArgumentList());
-                        continue;
+                        return _syntaxFactory.InvocationExpression(expr, this.ParseParenthesizedArgumentList());
 
+                    // a?.b[...]
                     case SyntaxKind.OpenBracketToken:
-                        expr = _syntaxFactory.ElementAccessExpression(expr, this.ParseBracketedArgumentList());
-                        continue;
+                        return _syntaxFactory.ElementAccessExpression(expr, this.ParseBracketedArgumentList());
 
+                    // a?.b.
                     case SyntaxKind.DotToken:
-                        expr = _syntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expr, this.EatToken(), this.ParseSimpleName(NameOptions.InExpression));
-                        continue;
+                        return _syntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expr, this.EatToken(), this.ParseSimpleName(NameOptions.InExpression));
 
-                    case SyntaxKind.QuestionToken:
-                        return !CanStartConsequenceExpression()
-                            ? expr
-                            : _syntaxFactory.ConditionalAccessExpression(
-                                expr,
-                                operatorToken: this.EatToken(),
-                                ParseConsequenceSyntax());
-
-                    default:
-                        return expr;
+                    // a?.b?
+                    case SyntaxKind.QuestionToken when CanStartConsequenceExpression():
+                        return _syntaxFactory.ConditionalAccessExpression(
+                            expr,
+                            operatorToken: this.EatToken(),
+                            ParseConsequenceSyntax());
                 }
-            }
 
-            bool isOptionalExclamationsFollowedByConditionalOperation()
-            {
-                var index = 0;
-                while (this.PeekToken(index).Kind == SyntaxKind.ExclamationToken)
-                    index++;
+                // a?.b = c
+                var (operatorTokenKind, operatorExpressionKind) = GetExpressionOperatorTokenKindAndExpressionKind();
+                if (IsExpectedAssignmentOperator(operatorTokenKind))
```

Comment:
unclear to me if precedence and associativity is correct here.   do you have tests for things like `a?.b = c = d`?  and if that is done properly?

also, this is in a while loop.  so i def want to understand if `a?.b = c ...` is correct if these are followed by `(`, `[`, `.`, and `?`.

My guess is that for the first three things are fine, since parsing the 'c' portion will consume those.  But i'm def unclear on what happens if the rhs has `?` stuff.  Can we have lots of tests around this, especially validating precedence/associativity?

`a?.b = c?.d = e` etc.

Reply:
> do you have tests for things like `a?.b = c = d`? and if that is done properly?

Yes.

> Can we have lots of tests around this, especially validating precedence/associativity?
>
> `a?.b = c?.d = e` etc.

Sure. The precedence and associativity simply comes from the assignment operator. So "assignment" precedence and right-associativity for nested assignments.